hera_pspec/uvwindow.py

```diff
@@ -0,0 +1,946 @@
+            else:  
+                self.ft_file = ftbeam
+        else:
+            raise ValueError('Wrong ftbeam input. See docstring.')
+        self.mapsize = None # Size of the flat map the beam was projected onto. 
+                            # Only used for internal calculations.
+
+        # if data file is used, initialises related arguments.   
+        if isinstance(uvdata, str):
+            if (uvdata == ''):
+                self.is_uvdata = False  
+            else: 
+                self.is_uvdata = True
+                self.uvdata = UVData()
+                self.uvdata.read(uvdata, read_data=False)
+        else:
+            self.is_uvdata = True
+            self.uvdata = uvdata
+
+        # Analysis-related attributes. 
+        # Will be set with set_spw_range andd set_spw_parameters
+        # once one of the get_wf functions is called.
+        self.freq_array = None
+        self.Nfreqs = None
+        self.dly_array = None
+        self.spw_range = None
+        self.avg_nu = None
+        self.avg_z = None
+        self.pol = None
+
+        # Initialise empty arrays for values potentially
+        # stored later (see get_spherical_wf)
+        self.kbins = []
+        self.kpara_bins = []
+        self.kperp_bins = []
+        self.cyl_wf = []
+        self.bl_lens = []
+        self.bl_weights = []
+
+    def get_bandwidth(self,file):
+        """
+        Read FT file to extract bandwidth it was computed along.
+
+        Parameters
+        ----------
+        file : str
+            Path to FT beam file.
+            Root name of the file to use, without the polarisation
+                Ex : FT_beam_HERA_dipole (+ path)
+            If '', then the object ft_file attribute is used.
+
+        Returns
+        ----------
+        bandwidth : array of floats
+            List of frequencies covered by the instrument, in Hz.
+        """
+
+        assert self.pol is not None, "Need to set polarisation first."
+
+        filename = '%s_%s.hdf5' %(file,self.pol)
+        assert os.path.isfile(filename), "Cannot find FT beam file: %s." %filename
+        f = h5py.File(filename, "r") 
+        HERA_bw = f['freq'][...]
+        f.close()
+
+        assert len(HERA_bw)>1, "Error reading file, empty bandwidth."
+
+        return HERA_bw
+
+    def set_taper(self, taper, clear_cache = True):
+        """
+        Set data tapering type.
+
+        Parameters
+        ----------
+        taper : str
+            Type of data tapering. See uvtools.dspec.gen_window for options.
+        clear_cache : bool, optional
+            Clear saved window functions if existing
+            (in case they were computed with another taper)
+        """
+        self.taper = taper
+        if (len(self.cyl_wf)>0):
+            raise_warning('New taper set but window functions have not been updated.')
+            if clear_cache:
+                self.clear_cache(clear_cyl_bins=True)
+
+
+    def set_spw_range(self,spw_range):
```

Comment:
I've modified the class so that `spw_range` is now an attribute of the `UVWindow` object, that needs to be defined when initialising the object as @steven-murray was pointing out that redefining  the option allowing the user to change the spectral window of the object mid-course might be confusing to the user. The point of specifying the spectral window is that you do not have to load the array containing the FT of the beam on the whole bandwidth to do your computations (which is pretty heavy), but just a portion of it.
From what I've seen most of the `hera_pspec` code works by looping over `spw`, then `pol`, then time,... so I don't think we can avoid re-defining a new `UVWindow` object at each iteration of these loops as the FT_beam file loaded depends on the polarisation. In the new version of the code, it would be only one line of code. Does that seem reasonable to you?